src/Symfony/Component/DependencyInjection/Compiler/CheckTypeHintsPass.php

```diff
@@ -0,0 +1,184 @@
+
+        if (count($configurationArguments) < $numberOfRequiredParameters) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid definition for service "%s": "%s::%s()" requires %d arguments, %d passed.', $this->currentId, $reflectionFunction->class, $reflectionFunction->name, $numberOfRequiredParameters, count($configurationArguments)));
+        }
+
+        $reflectionParameters = $reflectionFunction->getParameters();
+        $checksCount = min($reflectionFunction->getNumberOfParameters(), count($configurationArguments));
+
+        for ($i = 0; $i < $checksCount; ++$i) {
+            if (!$reflectionParameters[$i]->hasType() || $reflectionParameters[$i]->isVariadic()) {
+                continue;
+            }
+
+            $this->checkTypeHint($configurationArguments[$i], $reflectionParameters[$i]);
+        }
+
+        if ($reflectionFunction->isVariadic() && ($lastParameter = end($reflectionParameters))->hasType()) {
+            $variadicParameters = array_slice($configurationArguments, $lastParameter->getPosition());
+
+            foreach ($variadicParameters as $variadicParameter) {
+                $this->checkTypeHint($variadicParameter, $lastParameter);
+            }
+        }
+    }
+
+    /**
+     * Check type hints compatibility between
+     * a definition argument and a reflection parameter.
+     *
+     * @throws InvalidArgumentException on type hint incompatibility
+     */
+    private function checkTypeHint($configurationArgument, \ReflectionParameter $parameter): void
+    {
+        $referencedDefinition = $configurationArgument;
+
+        if ($referencedDefinition instanceof Reference) {
+            $referencedDefinition = $this->container->findDefinition((string) $referencedDefinition);
+        }
+
+        if ($referencedDefinition instanceof Definition) {
+            $class = $this->getClassName($referencedDefinition);
+
+            if (!$this->autoload && !class_exists($class, false)) {
+                return;
+            }
+
+            if (!is_a($class, $parameter->getType()->getName(), true)) {
+                throw new InvalidParameterTypeHintException($this->currentId, null === $class ? 'null' : $class, $parameter);
+            }
+        } else {
+            if (null === $configurationArgument && $parameter->allowsNull()) {
+                return;
+            }
+
+            if ($parameter->getType()->isBuiltin() && is_scalar($configurationArgument)) {
+                return;
+            }
+
+            if ('iterable' === $parameter->getType()->getName() && $configurationArgument instanceof IteratorArgument) {
+                return;
+            }
+
+            if ('Traversable' === $parameter->getType()->getName() && $configurationArgument instanceof IteratorArgument) {
+                return;
+            }
+
+            $checkFunction = 'is_'.$parameter->getType()->getName();
```

Comment:
`sprintf('is_%s', $parameter->getType()->getName())` can be used here. Looks more neat than appending